File: services/ingestion-A2/app/handler.py
from __future__ import annotations

import os
from typing import Any

import boto3

from .bedrock import TerminalExtractionError, extract_receipt
from .key import parse_document_key
from .redaction import redact_extraction
from .repository import ExtractionRepository


def _required(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        raise RuntimeError(f"Missing required environment variable: {name}")
    return value


def handler(event: dict[str, Any], _: Any) -> None:
    region = _required("AWS_REGION")
    model_id = _required("BEDROCK_MODEL_ID")
    repository = ExtractionRepository(_required("COCKROACH_DATABASE_URL"))
    s3 = boto3.client("s3", region_name=region)

    for record in event["Records"]:
        identity = parse_document_key(record["s3"]["object"]["key"])
        try:
            if model_id == "DISABLED":
                raise TerminalExtractionError("Bedrock extraction is disabled by the Free Tier cost guard")
            response = s3.get_object(Bucket=record["s3"]["bucket"]["name"], Key=identity.s3_key)
            content_type = response.get("ContentType")
            extraction = extract_receipt(region, model_id, response["Body"].read(), content_type)
            repository.save_extraction(identity, redact_extraction(extraction))
        except TerminalExtractionError as error:
            repository.mark_needs_review(identity, str(error))

File: services/ingestion/s3-extraction/src/handler.ts
// Lambda entry point (S3 ObjectCreated trigger). Port of app/handler.py.
// Minimal local event type instead of a full @types/aws-lambda dependency
// — only the two fields actually used are typed.

import { GetObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { parseDocumentKey } from './key';
import { extractReceipt, redactExtraction, TerminalExtractionError } from './extract';
import { saveExtraction, markNeedsReview } from './repository';
import { IS_MOCK } from './client';
import type { ContentType } from './types';

interface S3EventRecord {
  s3: {
    bucket: { name: string };
    object: { key: string };
  };
}

export interface S3Event {
  Records: S3EventRecord[];
}

function required(name: string): string {
  const value = process.env[name];
  if (!value) throw new Error(`Missing required environment variable: ${name}`);
  return value;
}

async function streamToBytes(stream: AsyncIterable<Uint8Array> | undefined): Promise<Uint8Array> {
  if (!stream) return new Uint8Array();
  const chunks: Uint8Array[] = [];
  for await (const chunk of stream) chunks.push(chunk);
  return Buffer.concat(chunks);
}

export async function handler(event: S3Event): Promise<void> {
  // A deployed Lambda invocation must always have COCKROACH_DATABASE_URL
  // set — IS_MOCK falling back to the local fixture file here would look
  // like success while silently writing nothing to the real database.
  // Mock mode is a CLI-only convenience (see cli.ts).
  if (IS_MOCK) {
    throw new Error('Missing required environment variable: COCKROACH_DATABASE_URL');
  }

  const region = required('AWS_REGION');
  const modelId = required('BEDROCK_MODEL_ID');
  const s3 = new S3Client({ region });

  for (const record of event.Records) {
    const identity = parseDocumentKey(record.s3.object.key);
    try {
      if (modelId === 'DISABLED') {
        throw new TerminalExtractionError('Bedrock extraction is disabled by the Free Tier cost guard');
      }
      const response = await s3.send(new GetObjectCommand({ Bucket: record.s3.bucket.name, Key: identity.s3Key }));
      const contentType = response.ContentType as ContentType;
      const body = await streamToBytes(response.Body as AsyncIterable<Uint8Array> | undefined);
      const extraction = await extractReceipt(region, modelId, body, contentType);
      await saveExtraction(identity, redactExtraction(extraction));
    } catch (err) {
      if (err instanceof TerminalExtractionError) {
        await markNeedsReview(identity, err.message);
      } else {
        throw err;
      }
    }
  }
}
